### PR TITLE
[Monitor] Fixing issue #3957: bug in Get-AzureRmLog cmdlet, returning…

### PR DESCRIPTION
**Pull request by gucalder**
#4056 

## Description
- Fixing issue #3957: bug in Get-AzureRmLog cmdlet, returning at most 200 records despite of MaxEvents.
- Adding an alias to the MaxEvents parameter, i.e.: MaxRecords.
- Fixing documentation for this command.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Insights/Commands.Insights/ActivityLogs/GetAzureRmLogCommand.cs [hunk 1]
```diff
@@ -90,11 +90,12 @@ public class GetAzureRmLogCommand : LogsCmdletBase
...
         /// <summary>
-        /// Gets or sets the max number of events to fetch parameter of the cmdlet
+        /// Gets or sets the max number of records to fetch parameter of the cmdlet
         /// </summary>
-        [Parameter(ValueFromPipelineByPropertyName = true, HelpMessage = "The maximum number of events to fetch")]
+        [Parameter(ValueFromPipelineByPropertyName = true, HelpMessage = "The maximum number of records to fetch. Alias: MaxEvents for backwards compatibility")]
+        [Alias("MaxEvents")]
         [ValidateNotNullOrEmpty]
-        public virtual int MaxEvents { get; set; }
+        public virtual int MaxRecords { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/ActivityLogs/GetAzureRmLogCommand.cs]: @gucalder any reason for the parameter name change?

**Comment by cormacpayne** [comment]
@gucalder Hey Guillermo, is this PR ready for review?

**Comment by gucalder** [comment]
Now it should be ready for review.

File: src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs [hunk 2]
```diff
@@ -105,19 +105,17 @@ protected virtual TimeSpan GetDefaultQueryTimeRange()
...
         /// </summary>
-        protected virtual void SetMaxEventsIfPresent(string currentQueryFilter, string name, int value)
+        protected virtual void SetMaxEventsIfPresent(string currentQueryFilter, int value)
         {
-            if (value > 0 && value <= 100000)
-            {
-                this.MaxEvents = value;
-            }
+            // If value is not acceptable this forces the use of the default value
+            this.MaxRecords = (value > 0 && value <= 100000) ? value : 0;
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs]: @gucalder in the change log, you mentioned 

_Any value for MaxEvents that is less than 1 is ignored and the default is used instead_

and that the default value is 1000. However, here you are setting the value to 0. If this is setting something different, ignore this comment 😄

File: src/ResourceManager/Insights/Commands.Insights.Test/ScenarioTests/ActivityLogsTests.ps1 [hunk 3]
```diff
@@ -120,7 +120,71 @@ function Test-GetAzureSubscriptionIdLog
...
+        # Cleanup
+        # No cleanup needed for now
+    }
+}
+
+<#
+.SYNOPSIS
+Tests getting the logs for a subscription Id.
+#>
+function Test-GetAzureSubscriptionIdLogMaxEvents
```
Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights.Test/ScenarioTests/ActivityLogsTests.ps1]: @gucalder are these the tests containing the pages of size 10 that return 15 elements? (the test you mentioned in the email)

Comment (gucalder) [inline src/ResourceManager/Insights/Commands.Insights.Test/ScenarioTests/ActivityLogsTests.ps1]: Yes, that is one of the places where this is tested.
This is another scenario test for that. NOTE: the page here is set to 6 records, but the are 8 records in total.
**Test-GetAzureSubscriptionIdLogPaged** (same file as Test-GetAzureSubscriptionIdLogMaxEvents)

This is another place, the unit test below. Here the first page is 10 records and the second 5.
**GetAzureSubscriptionIdLogCommandParametersProcessing()**

I added more comments to clarify the goal of the individual tests and also added more cases to the existing scenario tests.

File: src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs (hunk 2 above)
Comment (gucalder) [inline src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs]: I know it is odd. 
There are historic reasons to do it that way. They have to do with the fact that this part of the code is shared by two other cmdlets where MaxEvents is not defined as a parameter, and also MaxEvents was added later.
Anyway this is one thing that I might change later to make it a bit less confusing.

Comment (cormacpayne) [inline src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs]: @gucalder just to make sure I understand, does the `MaxEvent` value get set to 1000 somewhere else (by default), or is the description in the change log incorrect?

Comment (gucalder) [inline src/ResourceManager/Insights/Commands.Insights/LogsCmdletBase.cs]: Yes, MaxEvent is assigned 1000 somewhere else.

**Pull request merged**